### DISCUSSION TITLE
Problem with starting MSSQL container

### DISCUSSION THREAD

**Discussion by kfrajtak** [testcontainers/testcontainers-dotnet#533] [General]
I am trying to connect to a MSSQL container with `tmpfs` mapped volumes and automatically assigned port using ADO.NET.

I built the container in advance from following Dockerfile using `docker build -t mytest/sqlserver.testcontainer -f Dockerfile .` command:
```
FROM mcr.microsoft.com/mssql/server:2022-latest

ENV ACCEPT_EULA=Y
ENV SA_PASSWORD=Password123
ENV LD_PRELOAD=/nodirect_open.so

USER root
RUN wget https://raw.githubusercontent.com/t-oster/mssql-docker-zfs/master/nodirect_open.c
RUN apt update
RUN apt install -y gcc
RUN gcc -shared -fpic -o /nodirect_open.so nodirect_open.c -ldl
RUN apt purge -y gcc
RUN apt clean
RUN echo "/nodirect_open.so" >> /etc/ld.so.preload

HEALTHCHECK --interval=1s --timeout=2s --start-period=0s --retries=10 \
CMD /opt/mssql-tools/bin/sqlcmd -S . -U sa -P Password123 -Q "SELECT \"READY\"" | grep -q "READY"
````

The containers starts, I have to wait a little bit for it to spin up, the connection string is set:

```
var password = "Password123";

_container = new TestcontainersBuilder<TestcontainersContainer>()
    .WithAutoRemove(true)
    .WithImage("mytest/sqlserver.testcontainer:latest")
    .WithPortBinding(1433, true) // port is automatically assigned
    .WithName("testcontainer-databasefixture-" + Guid.NewGuid().ToString().Replace("-", ""))
    .WithTmpfsMount("/var/opt/mssql/data")
    .WithTmpfsMount("/var/opt/mssql/log")
    .Build();

await _container.StartAsync();

_connectionString = $"Server={_container.Hostname},{_container.GetMappedPublicPort(1433)};Initial catalog=x;User Id=sa;Password={password};";
```

... but then it fails to connect because (I am not even able to connect to the container from SQL Server Management Studio)

> A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections. (provider: TCP Provider, error: 0 - No connection could be made because the target machine actively refused it.

However, when I change the port binding to `.WithPortBinding(14333, 1433)` it works 

And when I change it to a prevously used port like `.WithPortBinding(49172, 1433)` it does not work because 
> Docker API responded with status code=InternalServerError, response=Ports are not available: exposing port TCP 0.0.0.0:49172 -> 0.0.0.0:0: listen tcp 0.0.0.0:49172: bind: An attempt was made to access a socket in a way forbidden by its access permissions.)

but no application is using that port.

Can anybody help? Thanks

**Comment by HofmeisterAn** [comment]
Is the port `49172` in the reserved port range (run `netsh interface ipv4 show excludedportrange protocol=tcp` to get the reserved range)? See: https://github.com/docker/for-win/issues/3171#issue-395479404.

**Reply by kfrajtak** [reply]
I can't tell right now, but why is the test container started with such port mapped?

**Reply by HofmeisterAn** [reply]
`.WithPortBinding(1433, true)` automatically assigns a random host port. It looks like there is a bug in the underlying Docker Desktop for Windows implementation. You can either do the recommended steps in the comment above, assign a fix port or something like: https://github.com/testcontainers/testcontainers-dotnet/issues/449#issuecomment-1165785770.

**Reply by kfrajtak** [reply]
Do you think a PR with checking allowed port range would fit your project?

**Reply by HofmeisterAn** [reply]
Did you try the latest version (2.3.0)? It contains a couple of improvments in regarding hostname and port resolution.
